File: comfyfleet/workflow.py
"""Loading, inspecting and patching API-format ComfyUI workflows."""

from __future__ import annotations

import copy
import json
import random
from pathlib import Path
from typing import Any, Iterator

from .config import Override

SEED_FIELDS = ("seed", "noise_seed", "rand_seed")

ASSET_FIELDS = ("image", "images", "video", "audio", "file", "mask", "filename", "path", "video_file")
ASSET_EXTENSIONS = {
    ".png", ".jpg", ".jpeg", ".webp", ".bmp", ".tif", ".tiff", ".gif",
    ".mp4", ".mov", ".mkv", ".webm", ".avi",
    ".wav", ".mp3", ".flac", ".ogg", ".m4a",
}

MAX_SEED = 0xFFFFFFFFFFFFFFFF


class WorkflowError(Exception):
    pass


class Workflow:
    """An API-format workflow: {node_id: {"inputs": {...}, "class_type": "...", "_meta": {...}}}."""

    def __init__(self, data: dict[str, Any], source: Path | None = None) -> None:
        self.data = data
        self.source = source

    # ---------------------------------------------------------------- loading

    @classmethod
    def load(cls, path: str | Path) -> "Workflow":
        path = Path(path)
        if not path.exists():
            raise WorkflowError(f"workflow not found: {path}")
        try:
            data = json.loads(path.read_text(encoding="utf-8"))
        except json.JSONDecodeError as exc:
            raise WorkflowError(f"{path} is not valid JSON: {exc}") from exc

        if isinstance(data, dict) and "prompt" in data and isinstance(data["prompt"], dict):
            data = data["prompt"]  # some tools wrap the API payload

        if not isinstance(data, dict):
            raise WorkflowError(f"{path}: expected a JSON object at the top level")

        if "nodes" in data and "links" in data:
            raise WorkflowError(
                f"{path} is a UI workflow, not an API workflow.\n"
                "  In ComfyUI: Workflow -> Export (API)  (enable Settings -> Lite Graph -> "
                "'Enable dev mode options' on older builds), then point the job at that file."
            )

        for node_id, node in data.items():
            if not isinstance(node, dict) or "class_type" not in node or "inputs" not in node:
                raise WorkflowError(
                    f"{path}: node {node_id!r} is not in API format "
                    "(expected 'class_type' and 'inputs'). Re-export with Export (API)."
                )
        return cls(data, source=path)

    def clone(self) -> "Workflow":
        return Workflow(copy.deepcopy(self.data), source=self.source)

    # -------------------------------------------------------------- inspection

    def nodes(self) -> Iterator[tuple[str, dict[str, Any]]]:
        yield from self.data.items()

    def class_types(self) -> set[str]:
        return {str(node.get("class_type")) for _, node in self.nodes()}

    @staticmethod
    def title_of(node: dict[str, Any]) -> str:
        meta = node.get("_meta") or {}
        return str(meta.get("title") or node.get("class_type") or "")

    def seed_nodes(self) -> list[tuple[str, str]]:
        """[(node_id, field)] for every literal seed widget."""
        found: list[tuple[str, str]] = []
        for node_id, node in self.nodes():
            for field_name, value in (node.get("inputs") or {}).items():
                if field_name in SEED_FIELDS and isinstance(value, (int, float)) and not isinstance(value, bool):
                    found.append((node_id, field_name))
        return found

    def asset_refs(self) -> list[tuple[str, str, str]]:
        """[(node_id, field, filename)] for widget values that look like input files."""
        found: list[tuple[str, str, str]] = []
        for node_id, node in self.nodes():
            for field_name, value in (node.get("inputs") or {}).items():
                if not isinstance(value, str) or not value:
                    continue
                suffix = Path(value).suffix.lower()
                if suffix in ASSET_EXTENSIONS and (
                    field_name in ASSET_FIELDS or "image" in field_name or "video" in field_name
                ):
                    found.append((node_id, field_name, value))
        return found

    def widget_values(self) -> list[tuple[str, str, str, str]]:
        """[(node_id, class_type, field, value)] for every string widget - used for model checks."""
        found: list[tuple[str, str, str, str]] = []
        for node_id, node in self.nodes():
            class_type = str(node.get("class_type"))
            for field_name, value in (node.get("inputs") or {}).items():
                if isinstance(value, str):
                    found.append((node_id, class_type, field_name, value))
        return found

    # ---------------------------------------------------------------- patching

    def apply_overrides(self, overrides: list[Override]) -> list[str]:
        """Mutates in place. Returns human-readable descriptions of what changed."""
        applied: list[str] = []
        for override in overrides:
            targets = list(self._match_nodes(override))
            if not targets:
                raise WorkflowError(f"override {override.describe()} matched no node")
            for node_id, node in targets:
                inputs = node.setdefault("inputs", {})
                if override.field_name not in inputs:
                    raise WorkflowError(
                        f"override {override.describe()}: node {node_id} "
                        f"({node.get('class_type')}) has no input {override.field_name!r}. "
                        f"Available: {', '.join(sorted(inputs))}"
                    )
                if isinstance(inputs[override.field_name], list):
                    raise WorkflowError(
                        f"override {override.describe()}: node {node_id}.{override.field_name} is "
                        "wired to another node, not a literal widget - it cannot be overridden."
                    )
                inputs[override.field_name] = override.value
                preview = str(override.value)
                if len(preview) > 60:
                    preview = preview[:57] + "..."
                applied.append(f"{node_id}.{override.field_name} = {preview}")
        return applied

    def _match_nodes(self, override: Override) -> Iterator[tuple[str, dict[str, Any]]]:
        for node_id, node in self.nodes():
            if override.node_id is not None and node_id == override.node_id:
                yield node_id, node
            elif override.title is not None and self.title_of(node) == override.title:
                yield node_id, node
            elif override.class_type is not None and node.get("class_type") == override.class_type:
                yield node_id, node

    def set_seed(self, seed: int) -> int:
        """Set every literal seed widget. Returns the number of widgets patched."""
        count = 0
        for node_id, field_name in self.seed_nodes():
            self.data[node_id]["inputs"][field_name] = int(seed) & MAX_SEED
            count += 1
        return count

    def rewrite_asset_names(self, mapping: dict[str, str]) -> None:
        """Point asset widgets at the names the remote machine returned after upload."""
        for node_id, field_name, value in self.asset_refs():
            if value in mapping and mapping[value] != value:
                self.data[node_id]["inputs"][field_name] = mapping[value]


def random_seed() -> int:
    return random.randint(0, MAX_SEED)

File: src/workflow.js
/** Loading, inspecting and patching API-format ComfyUI workflows. */

import fs from 'node:fs';
import path from 'node:path';

export class WorkflowError extends Error {}

const SEED_FIELDS = ['seed', 'noise_seed', 'rand_seed'];

const ASSET_FIELDS = ['image', 'images', 'video', 'audio', 'file', 'mask', 'filename', 'path', 'video_file'];
const ASSET_EXTENSIONS = new Set([
  '.png', '.jpg', '.jpeg', '.webp', '.bmp', '.tif', '.tiff', '.gif',
  '.mp4', '.mov', '.mkv', '.webm', '.avi',
  '.wav', '.mp3', '.flac', '.ogg', '.m4a',
]);

// Seeds stay inside the exact-integer range so a round trip through JSON never drifts.
const MAX_SEED = Number.MAX_SAFE_INTEGER;

export class Workflow {
  constructor(data, source = null) {
    this.data = data;
    this.source = source;
  }

  static parse(data, source = null) {
    if (data && typeof data === 'object' && data.prompt && typeof data.prompt === 'object') {
      data = data.prompt; // some tools wrap the API payload
    }
    if (!data || typeof data !== 'object' || Array.isArray(data)) {
      throw new WorkflowError('expected a JSON object at the top level');
    }
    if (data.nodes && data.links) {
      throw new WorkflowError(
        'This is a UI workflow, not an API workflow. In ComfyUI use Workflow -> Export (API) ' +
          'and select that file instead.',
      );
    }
    for (const [nodeId, node] of Object.entries(data)) {
      if (!node || typeof node !== 'object' || !node.class_type || !node.inputs) {
        throw new WorkflowError(
          `node "${nodeId}" is not in API format (expected "class_type" and "inputs"). ` +
            'Re-export with Workflow -> Export (API).',
        );
      }
    }
    return new Workflow(data, source);
  }

  static load(file) {
    if (!fs.existsSync(file)) throw new WorkflowError(`workflow not found: ${file}`);
    let parsed;
    try {
      parsed = JSON.parse(fs.readFileSync(file, 'utf8'));
    } catch (err) {
      throw new WorkflowError(`${path.basename(file)} is not valid JSON: ${err.message}`);
    }
    return Workflow.parse(parsed, file);
  }

  clone() {
    return new Workflow(structuredClone(this.data), this.source);
  }

  nodes() {
    return Object.entries(this.data);
  }

  classTypes() {
    return new Set(this.nodes().map(([, node]) => String(node.class_type)));
  }

  static titleOf(node) {
    return String(node?._meta?.title || node?.class_type || '');
  }

  /** [{nodeId, field}] for every literal seed widget. */
  seedNodes() {
    const found = [];
    for (const [nodeId, node] of this.nodes()) {
      for (const [field, value] of Object.entries(node.inputs || {})) {
        if (SEED_FIELDS.includes(field) && typeof value === 'number') found.push({ nodeId, field });
      }
    }
    return found;
  }

  /** [{nodeId, field, value}] for widget values that look like input files. */
  assetRefs() {
    const found = [];
    for (const [nodeId, node] of this.nodes()) {
      for (const [field, value] of Object.entries(node.inputs || {})) {
        if (typeof value !== 'string' || !value) continue;
        const ext = path.extname(value).toLowerCase();
        const looksLikeAsset = ASSET_FIELDS.includes(field) || field.includes('image') || field.includes('video');
        if (ASSET_EXTENSIONS.has(ext) && looksLikeAsset) found.push({ nodeId, field, value });
      }
    }
    return found;
  }

  /** [{nodeId, classType, field, value}] for every string widget - used for model checks. */
  widgetValues() {
    const found = [];
    for (const [nodeId, node] of this.nodes()) {
      for (const [field, value] of Object.entries(node.inputs || {})) {
        if (typeof value === 'string') found.push({ nodeId, classType: String(node.class_type), field, value });
      }
    }
    return found;
  }

  /** A flat description of the graph for the browser's node inspector. */
  outline() {
    return this.nodes()
      .map(([nodeId, node]) => ({
        id: nodeId,
        title: Workflow.titleOf(node),
        classType: String(node.class_type),
        widgets: Object.entries(node.inputs || {})
          .filter(([, value]) => !Array.isArray(value)) // arrays are wires to other nodes
          .map(([field, value]) => ({ field, value })),
      }))
      .sort((a, b) => {
        const na = Number(a.id);
        const nb = Number(b.id);
        if (Number.isFinite(na) && Number.isFinite(nb)) return na - nb;
        return String(a.id).localeCompare(String(b.id));
      });
  }

  /* -------------------------------------------------------- patching */

  matchNodes(override) {
    return this.nodes().filter(([nodeId, node]) => {
      if (override.node) return nodeId === String(override.node);
      if (override.title) return Workflow.titleOf(node) === override.title;
      if (override.class) return node.class_type === override.class;
      return false;
    });
  }

  /** Mutates in place. Returns human-readable descriptions of what changed. */
  applyOverrides(overrides) {
    const applied = [];
    for (const override of overrides || []) {
      const targets = this.matchNodes(override);
      const label = override.node || override.title || override.class;
      if (!targets.length) throw new WorkflowError(`override ${label}.${override.field} matched no node`);
      for (const [nodeId, node] of targets) {
        const inputs = node.inputs || (node.inputs = {});
        if (!(override.field in inputs)) {
          throw new WorkflowError(
            `override ${label}.${override.field}: node ${nodeId} (${node.class_type}) has no input ` +
              `"${override.field}". Available: ${Object.keys(inputs).sort().join(', ')}`,
          );
        }
        if (Array.isArray(inputs[override.field])) {
          throw new WorkflowError(
            `override ${label}.${override.field}: node ${nodeId}.${override.field} is wired to another ` +
              'node, not a literal widget - it cannot be overridden.',
          );
        }
        inputs[override.field] = override.value;
        const preview = String(override.value);
        applied.push(`${nodeId}.${override.field} = ${preview.length > 60 ? `${preview.slice(0, 57)}...` : preview}`);
      }
    }
    return applied;
  }

  /** Set every literal seed widget. Returns how many were patched. */
  setSeed(seed) {
    let count = 0;
    for (const { nodeId, field } of this.seedNodes()) {
      this.data[nodeId].inputs[field] = Number(seed) % (MAX_SEED + 1);
      count += 1;
    }
    return count;
  }

  /** Point asset widgets at the names the remote machine returned after upload. */
  rewriteAssetNames(mapping) {
    for (const { nodeId, field, value } of this.assetRefs()) {
      const mapped = mapping[value] ?? mapping[path.basename(value)];
      if (mapped && mapped !== value) this.data[nodeId].inputs[field] = mapped;
    }
  }
}

export function randomSeed() {
  return Math.floor(Math.random() * MAX_SEED);
}

export { MAX_SEED };
